src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -237,51 +237,136 @@ static const Type* bitshuffle_value(const TypeInteger* src_type, const TypeInteg
 
   jlong hi = bt == T_INT ? max_jint : max_jlong;
   jlong lo = bt == T_INT ? min_jint : min_jlong;
-
-  if(mask_type->is_con() && mask_type->get_con_as_long(bt) != -1L) {
+  assert(bt == T_INT || bt == T_LONG, "");
+
+  // Rule 1: Bit compression selects the source bits corresponding to true mask bits,
+  // packs them and places them contiguously at destination bit positions
+  // starting from least significant bit, remaining higher order bits are set
+  // to zero.
+
+  // Rule 2: Bit expansion is a reverse process, which sequentially reads source bits
+  // starting from LSB and places them at bit positions in result value where
+  // corresponding mask bits are 1. Thus, bit expansion for non-negative mask
+  // value will always generate a +ve value, this is because sign bit of result
+  // will never be set to 1 as corresponding mask bit is always 0.
+
+  // Case A) Constant mask
+  if (mask_type->is_con()) {
+    // Case A.1 bit compression:-
+    // Result.Hi = popcount(1 << mask_bits - 1)
+    // Result.Lo = min iff mask == -1 assuming all source bits apart from most
+    //                                significant bit were set to 0
+    //       else
+    // Result.Lo = 0 iff atleast one mask bit is zero, corresponding source
+    // bit will be masked, hence result of bit compression will be a +ve
+    // value.
+    // e.g.
+    //  src = 0xXXXXXXXX (non-constant source)
+    //  mask = 0xEFFFFFFF (constant mask)
+    //  result.hi = 0x7FFFFFFF
+    //  result.lo = 0
     jlong maskcon = mask_type->get_con_as_long(bt);
-    int bitcount = population_count(static_cast<julong>(bt == T_INT ? maskcon & 0xFFFFFFFFL : maskcon));
     if (opc == Op_CompressBits) {
-      // Bit compression selects the source bits corresponding to true mask bits
-      // and lays them out contiguously at destination bit positions starting from
-      // LSB, remaining higher order bits are set to zero.
-      // Thus, it will always generate a +ve value i.e. sign bit set to 0 if
-      // any bit of constant mask value is zero.
-      lo = 0L;
-      hi = (1UL << bitcount) - 1;
+      int bitcount = population_count(static_cast<julong>(bt == T_INT ? maskcon & 0xFFFFFFFFL : maskcon));
+      hi = maskcon == -1L ? hi : (1UL << bitcount) - 1;
+      lo = maskcon == -1L ? lo : 0L;
```

Comment:
It could be nice to have a proper `if-else` here, and add the comments to each scope, rather than above. That would allow you to avoid duplicating the code above in the comments.